### ISSUE TITLE
Context and virtual components

### ISSUE THREAD

**Issue by orlov-vo** [matthewp/haunted#40]
Now we have a React-like context (see issue  #17). But we can't use it in an application built with virtual components only (without web components).

I think we should provide some directives for `lit-html` after creating context. However, we need to discuss and choose the comfortable api for it...

An abstract example of usage:

```js
const ThemeContext = createContext('dark');

const { provide, consume } = ThemeContext;

const MyConsumer = virtual(() => {
  const context = useContext(ThemeContext);

  return context;
});

const App = virtual(() => {
  const [theme, setTheme] = useState('light');

  return html`
      <select value=${theme} @change=${(e) => setTheme(e.target.value)}>
        <option value="dark">Dark</option>
        <option value="light">Light</option>
      </select>
      
      ${provide(theme, html`
        ${MyConsumer()}
        ${provide(theme === 'dark' ? 'light' : 'dark', html`
          ${consume((value) => html`<h1>${value}</h1>`)}
        `)}
      `)}
  `;
});

render(App(), document.body);
```

**Comment by matthewp** [comment]
@askbeka Can you remember what the problem was with context in virtual components?

**Comment by askbeka** [comment]
Yes, problem is with unsubscribing from context changes, since virtual components do not have a knowledge when they are removed, neither do parts. One of the way could be to put text or comment element and check if it still present in DOM in interval.

I am workings through couple of ideas, for DOM traversal, to find virtual provider, from virtual consumer, which are abstractions of collection of nodes they cover in templates.
But events are no longer beneficial in that case, it is better to just traverse manually, and it is faster too.

**Comment by matthewp** [comment]
Can we use a MutationObserver to know when it's removed?

**Comment by askbeka** [comment]
@matthewp Not really. since directive do not have their own parts, parts can be reused across directives.
consider:
```javascript
html`${predicate ? ComponentOne() : ComponentTwo()}`
```
they will use the same part and the same marker nodes, which means that when Component is no longer used markers will not be removed.

Exmaple above will not work with virtual components, I have addressed it in [this issue](https://github.com/matthewp/haunted/issues/42)

Only way to do this is to associate part with directive, so that when different directive is used with the same part, previous directive gets revoked.

Same should happend when previous part gets removed,
I will file an issue, future request to lit-html, see if they have some solution in mind

**Comment by djalmajr** [comment]
Any news about it? 😁

**Comment by djalmajr** [comment]
Polymer/lit-html#283

**Comment by ianldgs** [comment]
> [Polymer/lit-html#283](https://github.com/Polymer/lit-html/issues/283)

Issue is closed 🚀

**Comment by klauss194** [comment]
Why the docs still say that it can't be used then ?
